File: Leetcode 1283 Find the Smallest Divisor Given a Threshold [medium] [array]/Find the Smallest Divisor Given a Threshold [binary-search].js
/**
 * @param {number[]} nums
 * @param {number} threshold
 * @return {number}
 */
var smallestDivisor = function(nums, threshold) {
    let max = Math.max(...nums);
    
    let start = 1;
    let end = max;
    while (start + 1 < end) {
        let mid = start + Math.floor((end - start) / 2);
        if (check(mid, nums, threshold)) {
            end = mid;
        } else {
            start = mid;
        }
    }
    
    if (check(start, nums, threshold)) {
        return start;
    }
    return end;
};

let check = (div, nums, threshold) => {
    let result = 0;
    nums.forEach(num => result += (Math.ceil(num / div)));
    return result <= threshold;
}

File: Leetcode 1283 Find the Smallest Divisor Given a Threshold [medium] [array]/Find the Smallest Divisor Given a Threshold [binary-search].java
class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        int max = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > max) {
                max = nums[i];
            }
        }
        System.out.println(max);
        
        int start = 1;
        int end = max;
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (isWithinThreshold(nums, threshold, mid)) {
                end = mid;
            } else {
                start = mid;
            }
        }
        
        if (isWithinThreshold(nums, threshold, start)) {
            return start;
        }
        return end;
    }
    
    private boolean isWithinThreshold(int[] nums, int threshold, int divisor) {
        int result = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] % divisor == 0) {
                result += (nums[i] / divisor);
            } else {
                result += (nums[i] / divisor) + 1;
            }
            
        }
        return result <= threshold;
    }
}